sound/soc/sof/intel/hda-loader.c

```diff
@@ -345,6 +395,12 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
 	struct firmware stripped_firmware;
 	int ret, ret1, i;
 
+	if ((sdev->fw_ready.flags & SOF_IPC_INFO_IMR_RESTORE) &&
+	    !sdev->first_boot) {
+		dev_info(sdev->dev, "info: IMR restore supported, booting from IMR directly\n");
+		return hda_dsp_boot_imr(sdev);
```

Comment:
what happens if we override this IMR_RESTORE configuration. Can we still go ahead and proceed with the default path?

This should be an opt-in solution IMHO. Changing the flows will mechanically lead to bugs, so by default users who upgrade the firmware should not see a change on how the DSP is managed.

Reply:
> what happens if we override this IMR_RESTORE configuration. Can we still go ahead and proceed with the default path?

Yes, if the driver choose to re-downloading, even the IMR_RESTORE is declared support from the fw_ready, we will still go ahead with today's path.

> 
> This should be an opt-in solution IMHO. Changing the flows will mechanically lead to bugs, so by default users who upgrade the firmware should not see a change on how the DSP is managed.

Um, that's what I commented above, do you suggest a kconfig item or a module param here?

Reply:
@plbossart This can also fix issues, so not sure if this should be opt-in. If tests look good, we could just make this the default and offer an opt-out in case an issue is hit.

Reply:
So let me add a kconfig item and the check will be like this:
```c
if ((sdev->fw_ready.flags & SOF_IPC_INFO_IMR_RESTORE) &&
            !CONFIG_SND_SOC_SOF_DEBUG_DISABLE_IMR_RESTORE &&
	    !sdev->first_boot) {
```

Reply:
@keyonjie @plbossart I'd vote for a module option. If this causes a regression, user can easily add a modprobe.d option, but having to roll out a custom kernel for a specific system is too difficult. And we have to too many Kconfig options. If there are many, we should have a quirk table in driver to make the default sane for most systems (as the module option is not scalable either).

Reply:
agree with @kv2019i we already have sof_debug quirks, let's add one bit to disable IMR reload.